### PR TITLE
Implement parseUrl, parseHook

### PR DESCRIPTION
**Pull request by FenrirUnbound**
In conforming to screwdriver-cd/scm-base#11, this PR aims to implement as much of the backwards-compatible changes as possible. Specifically, it implements `parseUrl()` and `parseHook()`.

The functions `decorateUrl()`, `decorateCommit()`, and `decorateAuthor()` were omitted because screwdriver-cd/scm-base#11 introduces a change that replaces`scmUrl` with `scmUri`. Implementing this change is a backwards-incompatible update. By deferring the 3 methods, we prevent the code base from having a mix of `scmUrl` and `scmUri`, which would be incredibly confusing: from a human-readability stand-point, 'i' and 'l' are difficult to distinguish at a quick glance.

Helps contribute to #15

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -325,6 +332,76 @@ class GithubScm extends Scm {
...
+     * necessary data to execute a Screwdriver job with.
+     * @method parseHook
+     * @param  {Object}  webhookPayload  The webhook payload received from the
+     *                                   SCM service.
+     * @param  {Object}  payloadHeaders  The request headers associated with the
+     *                                   webhook payload
+     * @return {Object}                  A key-map of data related to the received
+     *                                   payload
+     */
+    parseHook(webhookPayload, payloadHeaders) {
```
Comment (d2lam) [inline index.js]: Headers is the first param, payload is second. According to base.

File: index.js [hunk 2]
```diff
@@ -325,6 +332,76 @@ class GithubScm extends Scm {
...
+        const type = payloadHeaders['x-github-event'];
+        const checkoutUrl = hoek.reach(webhookPayload, 'repository.ssh_url');
+
+        switch (type) {
+        case 'pull_request': {
+            let action = hoek.reach(webhookPayload, 'action');
+            const prNum = hoek.reach(webhookPayload, 'pull_request.number');
+
+            if (!SUPPORTED_PR_ACTIONS.includes(action)) {
+                action = 'closed';
```
Comment (d2lam) [inline index.js]: If not supported here I think it should throw an error.

File: index.js [hunk 3]
```diff
@@ -325,6 +332,76 @@ class GithubScm extends Scm {
...
+     * 'token' is required, since it is necessary to lookup the SCM ID by
+     * communicating with said SCM service.
+     * @method parseUrl
+     * @param  {Object} config         Config object
+     * @param  {String} config.scmUrl  The scmUrl to parse
+     * @param  {String} config.token   The token used to authenticate to the SCM service
+     * @return {Promise}               Resolves to an ID of 'serviceName:repoId:branchName'
+     */
+    parseUrl(config) {
+        const scmInfo = getInfo(config.scmUrl);
```
Comment (tkyi) [inline index.js]: `scmUrl` is now `checkoutUrl`
https://github.com/screwdriver-cd/scm-github/pull/14/files#diff-04c6e90faac2675aa89e2176d2eec7d8R58

File: index.js (hunk 2 above)
Comment (FenrirUnbound) [inline index.js]: That's the default now: https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/webhooks/github.js#L238

File: index.js [hunk 4]
```diff
@@ -325,6 +332,76 @@ class GithubScm extends Scm {
...
+                prNum,
+                prRef: `${checkoutUrl}#pull/${prNum}/merge`,
+                sha: hoek.reach(webhookPayload, 'pull_request.head.sha'),
+                type: 'pr',
+                username: hoek.reach(webhookPayload, 'pull_request.user.login')
+            };
+        }
+        case 'push':
+            return {
+                action: 'opened',
```
Comment (FenrirUnbound) [inline index.js]: 👍

File: index.js (hunk 3 above)
Comment (FenrirUnbound) [inline index.js]: 👍

File: index.js [hunk 5]
```diff
@@ -325,6 +332,76 @@ class GithubScm extends Scm {
...
+        case 'pull_request': {
+            let action = hoek.reach(webhookPayload, 'action');
+            const prNum = hoek.reach(webhookPayload, 'pull_request.number');
+
+            if (!SUPPORTED_PR_ACTIONS.includes(action)) {
+                action = 'closed';
+            }
+
+            return {
+                action,
```
Comment (tkyi) [inline index.js]: should there be some check here to make the action `synchronized` with a 'd'?

File: index.js [hunk 6]
```diff
@@ -325,6 +332,76 @@ class GithubScm extends Scm {
...
+     * necessary data to execute a Screwdriver job with.
+     * @method parseHook
+     * @param  {Object}  payloadHeaders  The request headers associated with the
+     *                                   webhook payload
+     * @param  {Object}  webhookPayload  The webhook payload received from the
+     *                                   SCM service.
+     * @return {Object}                  A key-map of data related to the received
+     *                                   payload
+     */
+    parseHook(payloadHeaders, webhookPayload) {
```
Comment (tkyi) [inline index.js]: `_parseHook`

File: index.js [hunk 7]
```diff
@@ -325,6 +332,76 @@ class GithubScm extends Scm {
...
+     *
+     * 'token' is required, since it is necessary to lookup the SCM ID by
+     * communicating with said SCM service.
+     * @method parseUrl
+     * @param  {Object} config              Config object
+     * @param  {String} config.checkoutUrl  The checkout URL to parse
+     * @param  {String} config.token        The token used to authenticate to the SCM service
+     * @return {Promise}                    Resolves to an ID of 'serviceName:repoId:branchName'
+     */
+    parseUrl(config) {
```
Comment (tkyi) [inline index.js]: `_parseUrl`

File: index.js (hunk 6 above)
Comment (FenrirUnbound) [inline index.js]: It's not extending `scm-base`

File: index.js (hunk 7 above)
Comment (FenrirUnbound) [inline index.js]: It's not extending `scm-base` v2

File: index.js (hunk 5 above)
Comment (FenrirUnbound) [inline index.js]: 👍

File: index.js (hunk 6 above)
Comment (tkyi) [inline index.js]: 👍

File: README.md [hunk 8]
```diff
@@ -49,6 +49,45 @@ The `getCommitSha` function will fetch the commit sha for a given branch on a re
...
+
+### parseHook
+Required parameters:
+
+| Parameter        | Type  |  Description |
+| :-------------   | :---- | :-------------|
+| headers        | Object | The request headers associated with the webhook payload |
+| payload        | Object | The webhook payload received from the SCM service |
+
+#### Expected Outcome
```
Comment (d2lam) [inline README.md]: This should return a promise actually

File: index.js [hunk 9]
```diff
@@ -325,6 +332,78 @@ class GithubScm extends Scm {
...
+
+        switch (type) {
+        case 'pull_request': {
+            let action = hoek.reach(webhookPayload, 'action');
+            const prNum = hoek.reach(webhookPayload, 'pull_request.number');
+
+            if (action === 'synchronize') {
+                action = 'synchronized';
+            } else if (!SUPPORTED_PR_ACTIONS.includes(action)) {
+                action = 'closed';
```
Comment (d2lam) [inline index.js]: If it's not supported, I think it should throw an error

File: index.js (hunk 2 above)
Comment (FenrirUnbound) [inline index.js]: Removed

**Comment by FenrirUnbound** [comment]
@d2lam still awaiting your review

File: index.js [hunk 10]
```diff
@@ -325,6 +327,76 @@ class GithubScm extends Scm {
...
+        switch (type) {
+        case 'pull_request': {
+            let action = hoek.reach(webhookPayload, 'action');
+            const prNum = hoek.reach(webhookPayload, 'pull_request.number');
+
+            if (action === 'synchronize') {
+                action = 'synchronized';
+            }
+
+            return {
```
Comment (d2lam) [inline index.js]: Return a promise

Comment (FenrirUnbound) [inline index.js]: 👍

File: README.md (hunk 8 above)
Comment (FenrirUnbound) [inline README.md]: Doc'd

**Pull request merged**